### PR TITLE
Move get_new_subdir from padertorch to paderbox

### PR DESCRIPTION
**Pull request by boeddeker**
Changes:
 - Rename get_new_folder to get_new_subdir
 - Add prefix and suffix and remove try_id
 - Add cli
 - Add id_naming option

The first three commits are minor changes.

The discussion for this functions started in https://github.com/fgnt/padertorch/pull/46

### CODE REVIEW

File: paderbox/io/audioread.py [hunk 1]
```diff
@@ -276,7 +277,7 @@ def recursive_load_audio(
     kwargs = locals().copy()
     path = kwargs.pop('path')
 
-    if isinstance(path, (tuple, list)):
+    if isinstance(path, (tuple, list, types.GeneratorType)):
```
Comment (thequilo) [inline paderbox/io/audioread.py:280]: Is this related to `get_new_subdir`?

File: paderbox/io/new_subdir.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        basedir,
+        *,
+        id_naming='index',
+        mkdir=True,
+        prefix=None,
+        suffix=None,
+        consider_mpi=False,
+        dry_run=False,
+):
+    """Determine a new non existant sub directory.
```
Comment (thequilo) [inline paderbox/io/new_subdir.py:39]: Typo: existent (maybe: non-existent)

File: paderbox/io/new_subdir.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+def get_new_subdir(
+        basedir,
+        *,
+        id_naming='index',
+        mkdir=True,
+        prefix=None,
+        suffix=None,
+        consider_mpi=False,
+        dry_run=False,
```
Comment (thequilo) [inline paderbox/io/new_subdir.py:37]: Add type annotations?

File: paderbox/io/new_subdir.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+                This option ensures, that only one program gets the '2' and the
+                remaining programs search for another free id.
+        prefix:
+            Optional prefix for the id. e.g.: '2' -> '{prefix}_2'
+        suffix:
+            Optional suffix for the id. e.g.: '2' -> '2_{suffix}'
+        consider_mpi:
+            If True, only search on one mpi process for the folder and
+            distribute the folder name.
+            When using mpi with this option disables the following can/will
```
Comment (thequilo) [inline paderbox/io/new_subdir.py:74]: disabled? If disables is correct, I don't get this sentence.

File: paderbox/io/new_subdir.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        if dry_run:
+            print(f'dry_run: "os.makedirs({basedir})"')
+            # ToDo: Make this working.
+            #       Will fail when calling os.listdir
+        else:
+            basedir.mkdir(parents=True)
+
+    if Path('/net') in basedir.parents:
+        # If nt filesystem, assert not in /net/home
+        assert Path('/net/home') not in basedir.parents, basedir
```
Comment (thequilo) [inline paderbox/io/new_subdir.py:120]: Do we want to have this hard-coded in here or flexible as a parameter for easier use in other environments?

File: paderbox/io/new_subdir.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+                dir_nrs += [0]
+                _id = max(dir_nrs) + 1
+                _id = f'{prefix_}{_id}{_suffix}'
+        elif id_naming == 'time':
+            if i != 0:
+                time.sleep(1)
+            _id = datetime.datetime.now().strftime('%Y-%m-%d-%H-%M-%S')
+            _id = f'{prefix_}{_id}{_suffix}'
+        elif callable(id_naming):
+            _id = id_naming()
```
Comment (thequilo) [inline paderbox/io/new_subdir.py:157]: Check if the ID is unique in the subfolder? This is very unlikely for our naming functions below, but this might not be the case for a user-supplied naming function.

File: paderbox/io/new_subdir.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+            suffix:
+            consider_mpi:
+            dry_run:
+
+        Returns:
+
+        """
+        # Add more cli mappings if necessary
+        if id_naming == 'adjective_color_animal':
+            id_naming = NameGenerator(('adjectives', 'colors', 'animals'))
```
Comment (thequilo) [inline paderbox/io/new_subdir.py:337]: Put this into `get_new_subdir`?

File: paderbox/io/audioread.py (hunk 1 above)
Comment (boeddeker) [reply paderbox/io/audioread.py:280]: No, I committed this some days ago and forgot to push.
Since this is a minor change, I kept it in this PR.

File: paderbox/io/new_subdir.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    """Determine a new non existant sub directory.
+
+    Features:
+     - With mkdir: Thread and process save.
+     - Different conventions for ID naming possible, default running index.
+     - MPI aware: Get the folder on one worker and distribute to others.
+
+    Args:
+        basedir:
+            The new folder will be inside this directory
```
Comment (jensheit) [inline paderbox/io/new_subdir.py:48]: The new subdir will be inside ..

File: paderbox/io/new_subdir.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+            distribute the folder name.
+            When using mpi with this option disables the following can/will
+            happen
+             - When mkdir is True every process will get another folder.
+               i.e. each process has a folder just for this process.
+             - When mkdir is False it is likely that all get the same folder,
+               but this is not guarantied. Never use mpi with consider_mpi and
+               mkdir disabled.
+        dry_run:
+            Similar to mkdir. When true, disable mkdir and print the folder
```
Comment (jensheit) [inline paderbox/io/new_subdir.py:82]: Why is this similar to mkdir?

Did you mean:
When true, disables mkdir and prints the folder name
?

File: paderbox/io/new_subdir.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+            basedir.mkdir(parents=True)
+
+    if Path('/net') in basedir.parents:
+        # If nt filesystem, assert not in /net/home
+        assert Path('/net/home') not in basedir.parents, basedir
+
+    prefix_ = f'{prefix}_' if prefix else ''
+    _suffix = f'_{suffix}' if suffix else ''
+
+    for i in range(200):
```
Comment (jensheit) [inline paderbox/io/new_subdir.py:125]: The 200 are a bit arbitrary, should we set a variable somewhere so that it can be changed more easily?

File: paderbox/io/new_subdir.py (hunk 5 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py:120]: I am also not a fan for hard-coded things, but this one in special:
 - We recommend our students to code in the home directory, while storing the experiment files on a different file system.
    - We do this, because the home directory has an expensive backup system that is useful for code (i.e. never lose your written code).
    - Many other frameworks (e.g. Kaldi, ESPnet, Asteroid, ...) have the code and experiment files in the same folder, so I expect this split is special for us.
      - So others often don't care about this split.
- The naming of `/net/home` is very special. I don't expect that anyone will get a problem with this line. 
   - When someone reports, that a name collision, I will erase this assert.

File: paderbox/io/new_subdir.py (hunk 6 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py:157]: I don't get your point. The code later will check if the folder exist.
If it is the case, it will call again `id_naming` to get another name.
This is done up to 100 times, before the code complains, that it cannot find a new name.

File: paderbox/io/new_subdir.py (hunk 7 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py:337]: I am not sure, if I should put this inside (Actually I was unsure if the cli should have this).
On the python side using `NameGenerator(('adjectives', 'colors', 'animals'))` is more verbose and it is obvious, hot to change the generation.

File: paderbox/io/new_subdir.py (hunk 6 above)
Comment (thequilo) [reply paderbox/io/new_subdir.py:157]: Ups, I didn't get that. Then it's fine!

File: paderbox/io/new_subdir.py (hunk 10 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py:125]: It is even more arbitrary, because later the code ends the loop with the check `i > 100`.
I don't think we have to expose this parameter.

When someone gets a problem with this number, we can change it.
From my observations, it fails fast enough, when something is wrong and launching 100 experiments at the same time is difficult (When you launch with a delay of 10 ms it is unlikely that you get more than 1 collision).

File: paderbox/io/new_subdir.py (hunk 2 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py]: Done

File: paderbox/io/new_subdir.py (hunk 3 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py]: Done.

File: paderbox/io/new_subdir.py (hunk 4 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py:74]: I changed the complete statement.

File: paderbox/io/new_subdir.py (hunk 8 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py]: done

File: paderbox/io/new_subdir.py (hunk 9 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py]: > Why is this similar to mkdir?

Deleted

>  Did you mean:
> When true, disables mkdir and prints the folder name
> ?

Done

File: paderbox/io/new_subdir.py (hunk 10 above)
Comment (boeddeker) [reply paderbox/io/new_subdir.py:126]: Just for documentation: The origin for this loop is https://github.com/IDSIA/sacred/pull/285.

File: paderbox/io/new_subdir.py (hunk 7 above)
Comment (thequilo) [reply paderbox/io/new_subdir.py:338]: okay

File: paderbox/io/new_subdir.py (hunk 4 above)
Comment (thequilo) [reply paderbox/io/new_subdir.py:74]: I still don't get this sentence

Comment (boeddeker) [reply paderbox/io/new_subdir.py]: Sorry, I changed the text below. Now I updated this line to
`When using mpi (and `consider_mpi is False`) the following can/will happen`
